### DISCUSSION TITLE
Bedrock Build-in JSON mode

### DISCUSSION THREAD

**Discussion by jledesma-opsguru** [spring-projects/spring-ai#959] [Q&A]
Hi everyone,

Is the inability to configure Bedrock (Anthropic3ChatOptions) to generate JSON output a Bedrock limitation? It would be helpful if we could configure it like it is in Azure or OpenAI:

https://docs.spring.io/spring-ai/reference/api/structured-output-converter.html#_build_in_json_mode

**Comment by kimnamu** [comment]
Thanks for writing this up back in the M1 days, and for pointing at the exact docs section so the gap between the OpenAI/Azure JSON mode and Bedrock was easy to see.

Short version: at the time this was an upstream limitation rather than a Spring AI one. Neither the Anthropic Messages API nor Bedrock had a JSON mode, so `Anthropic3ChatOptions` had nothing to expose and the prompt-based `BeanOutputConverter` was the only route. Bedrock added structured outputs to the Converse API in February 2026, and Spring AI now supports that natively in `spring-ai-bedrock-converse`: `BedrockChatOptions` implements `StructuredOutputChatOptions`, so both `AdvisorParams.ENABLE_NATIVE_STRUCTURED_OUTPUT` and `outputSchema(...)` work.

| | |
|---|---|
| Landed in | 2.0.0-M3 (commit 3003316, closes #5445). Not in the 1.0.x / 1.1.x lines |
| Version to use | 2.0.0-M5 or later (feb6fe2 fixed `outputSchema` being dropped when options are merged). Verified below on 2.0.1 |
| Code path | `BedrockChatOptions#outputSchema` → `BedrockProxyChatModel#buildOutputConfig` → Converse `outputConfig.textFormat` with `type=json_schema` |
| Old `Anthropic3ChatModel` | Removed in 1.0.0-M6 together with the other InvokeModel chat models (upgrade notes). Converse is the only Bedrock chat path now |
| Docs | Bedrock Converse page, "Structured Output" section |

Run on 2.0.1 against `us.anthropic.claude-haiku-4-5-20251001-v1:0` in us-east-1, with an SDK `ExecutionInterceptor` printing the outgoing `ConverseRequest.outputConfig()`:

```
== 1) ChatClient + AdvisorParams.ENABLE_NATIVE_STRUCTURED_OUTPUT ==
[wire] outputConfig=OutputConfig(TextFormat=OutputFormat(Type=json_schema, Structure=*** Sensitive Data Redacted ***))
[entity] ActorsFilms[actor=Tom Hanks, movies=[Forrest Gump, The Shawshank Redemption, Saving Private Ryan]]
== 2) BedrockChatOptions.outputSchema directly ==
[wire] outputConfig=OutputConfig(TextFormat=OutputFormat(Type=json_schema, Structure=*** Sensitive Data Redacted ***))
[raw] {"actor": "Tom Hanks", "movies": ["Forrest Gump", "The Shawshank Redemption", "Saving Private Ryan"]}
[finishReason] end_turn
== 3) no outputSchema (control) ==
[wire] outputConfig=null
[raw] # Tom Hanks\n\n1. **Forrest Gump** (1994)\n2. **Saving Private Ryan** (1998)\n3. **Toy Story** (1995) - voice of Woody
```

<details>
<summary>Code used</summary>

```java
record ActorsFilms(String actor, List<String> movies) {}

BedrockProxyChatModel model = BedrockProxyChatModel.builder()
    .bedrockRuntimeClient(client) // plain BedrockRuntimeClient, us-east-1
    .options(BedrockChatOptions.builder()
        .model("us.anthropic.claude-haiku-4-5-20251001-v1:0")
        .maxTokens(300)
        .build())
    .build();

// 1) ChatClient path
ActorsFilms af = ChatClient.create(model).prompt()
    .advisors(AdvisorParams.ENABLE_NATIVE_STRUCTURED_OUTPUT)
    .user("Give me a real actor and three of their movies.")
    .call()
    .entity(ActorsFilms.class);

// 2) explicit schema
String schema = """
    {"type":"object","properties":{"actor":{"type":"string"},
     "movies":{"type":"array","items":{"type":"string"}}},
     "required":["actor","movies"],"additionalProperties":false}""";
ChatResponse resp = model.call(new Prompt("Give me a real actor and three of their movies.",
    BedrockChatOptions.builder()
        .model("us.anthropic.claude-haiku-4-5-20251001-v1:0")
        .outputSchema(schema)
        .build()));
```

</details>

One caveat: AWS gates `outputConfig` per model (the Spring docs link the Bedrock supported-models table), and Haiku 4.5 is the only one I ran. For models that are not on that list, the prompt-based `BeanOutputConverter` described in the structured-output docs is still the way to go.
